File: Lab7/Lab7_C++/LinkedList.cpp
#include "LinkedList.h"
#include <iostream>

LinkedList::LinkedList() {
	_head = nullptr; // При создании списка - он пустой, следовательно верхний элемент имеет пустой указатель.
	_size = 0;
}

void LinkedList::Push(short data) {
	_head = new Node(data, _head); // При добавлении элемента в начало, мы создаём новый объект типа Node и делаем его верхним.
	_size++;
}

LinkedList::Node::Node(short data, Node* pNext = nullptr) {
	this->data = data; // Поле data хранит в себе значение указателя.
	this->pNext = pNext; // Поле pNext хранит в себе указатель на следующий элемент.
}

int LinkedList::Size() {
	return _size;
}

short& LinkedList::operator[](const int index) {
	int counter = 0;
	Node* current = this->_head;
	while (current != nullptr) {
		if (counter == index)
			return current->data;
		current = current->pNext;
		counter++;
	}
}

int LinkedList::ThreeDivided() {
	int counter = 0;
	Node* current = this->_head; // Первым элементом - выбираем самый верхний.
	while (current != nullptr) {
		counter = (current->data % 3 == 0) ? ++counter : counter;
		current = current->pNext; // Переходим к следующему указателю.
	}		
	return counter;
}

double LinkedList::Mean() {
	double sum = 0;
	Node* current = this->_head;
	while (current != nullptr) {
		sum += current->data;
		current = current->pNext;
	}
	return sum / this->_size;
}

void LinkedList::MoreThanMeanRemove() {
	double Mean = this->Mean();
	Node* current = this->_head;
	int counter = 0;
	while (current != nullptr) {
		if (current->data > Mean) {
			current = current->pNext;
			RemoveAt(counter);
			continue;
		}
		counter++;
		current = current->pNext;
	}
}

void LinkedList::RemoveAt(int index) {
	if (index == 0)
		this->PopFront();
	else {
		Node* previous = this->_head;
		for (auto i = 0; i < index - 1; i++) 
			previous = previous->pNext;		
		Node* temp = previous->pNext;
		previous->pNext = temp->pNext;
		delete temp;
		_size--;
	}
}

void LinkedList::PopFront() {
	Node* temp = this->_head; // Создаём временную переменную, которая хранит в себе указатель на верхний элемент.
	_head = _head->pNext; // Перемещаем указатель на верхний элемент на следующий элемент.
	delete temp; // Очищаем память, от начального верхнего элемента.
	_size--; // уменьшаем размер массива.
}

void LinkedList::Print() {
	std::cout << "Your list: ";
	for (int i = 0; i < this->_size; i++) {
		std::cout << this->operator[](i);
		if (i == this->_size- 1)
			std::cout << "\n";
		else
			std::cout << "->";
	}
}

File: Lab7/Lab7_CS/LinkedList.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Lab8_CS
{
    class LinkedList
    {
        private class Node
        {
            public int _data;
            public Node _next;
            public Node(short data, Node _next = null)
            {
                this._data = data;
                this._next = _next;
            }
        }
        private Node _head;
        public int Size { get; private set; }
        public LinkedList()
        {
            _head = null;
            Size = 0;
        }
        public void Push(short data) 
        {
            _head = new Node(data, _head);
            Size++;
        }
        public int ThreeDivided()
        {
            int counter = 0;
            Node current = this._head;
            while(current != null)
            {
                counter = (current._data % 3 == 0) ? ++counter : counter;
                current = current._next;
            }
            return counter;
        }
        public double Mean()
        {
            double sum = 0;
            Node current = this._head;
            while(current != null)
            {
                sum += current._data;
                current = current._next;
            }
            return sum / this.Size;
        }
        public void MoreThanMeanRemove()
        {
            double Mean = this.Mean();
            Node current = this._head;
            int counter = 0;
            while(current != null)
            {
                if(current._data > Mean)
                {
                    current = current._next;
                    RemoveAt(counter);
                    continue;
                }
                counter++;
                current = current._next;
            }
        }
        public void RemoveAt(int index)
        {
            if (index == 0)
                PopFront();
            else
            {
                Node previous = this._head;
                for (var i = 0; i < index - 1; i++)
                    previous = previous._next;
                previous = previous._next._next;
                Size--;
            }
                
        }
        public void PopFront()
        {
            _head = _head._next;
            --Size;
        }
        public void Print()
        {
            int counter = 0;
            Node current = this._head;
            Console.WriteLine("Your list: ");
            while(current != null) 
            {
                Console.Write(current._data + ((counter != this.Size - 1) ? " -> " : "\n"));
                current = current._next;
                counter++;
            }                
        }
    }
}
